Physics interpolation (2D)

rburing opened the pull request:
This is a port of @lawnjelly's

- https://github.com/godotengine/godot/pull/76252/

to the master branch. It also takes bits
- from https://github.com/godotengine/godot/commit/b287f970dd34d59b8ea9fe6150e76b1c5777423d (before the hierarchical culling in 3.x),
- from the preceding 3D PR https://github.com/godotengine/godot/pull/52846 as needed,
- and from the later fixup PRs.

TODO:

- [x] Resolve questions marked by `TODO` in the code
- [x] Check snapping code
- [x] TileMap
- [x] Documentation (and mention that this only works for 2D for now)

Left for a follow-up PR (if it is acceptable):

- CPUParticles2D

Thanks to @lawnjelly for the original PR.

- Fixes https://github.com/godotengine/godot/issues/64182

Testing welcome! Make sure to enable the project setting `physics/common/physics_interpolation` and set `physics/common/physics_jitter_fix` to `0.0` when testing.

This PR is sponsored by My Spare Time™

rburing committed:
Fixed Timestep Interpolation (2D)

Adds fixed timestep interpolation to the rendering server (2D only).
Switchable on and off with a project setting (default is off).

Co-authored-by: lawnjelly <lawnjelly@gmail.com>

lawnjelly commented on scene/main/node.h:
```diff
@@ -86,6 +86,12 @@ class Node : public Object {
 		FLAG_PROCESS_THREAD_MESSAGES_ALL = 3,
 	};
 
+	enum PhysicsInterpolationMode {
```
```
	enum PhysicsInterpolationMode : unsigned int {
```
When using enums in bitfields, always specify unsigned. Some compilers default to signed for enums, and this can overflow the bits.

See:
https://github.com/godotengine/godot/pull/61757

lawnjelly commented on servers/rendering/renderer_canvas_cull.cpp:
```diff
@@ -247,17 +248,24 @@ void RendererCanvasCull::_cull_canvas_item(Item *p_canvas_item, const Transform2
 	if (snapping_2d_transforms_to_pixel) {
-		xform.columns[2] = xform.columns[2].round();
+		final_xform.columns[2] = final_xform.columns[2].round();
 		parent_xform.columns[2] = parent_xform.columns[2].round();
```
Note: If you are doing snapping here, you may want the snapping to apply to the local object, but NOT to propagate to the children, so they can do their snapping independently. This may entail creating an extra intermediate e.g. `snapped_xform` and use this for rendering, while passing the unchanged `final_xform` to the children.

This certainly needs consideration because snapping is done differently in 3.x.

adamscott replied:
I tried multiple times snap the children independently (ask @markdibarry) but it always resulted with snapping issues. I think it's best to snap the parent as is.

lawnjelly commented:
Great work @rburing . This is something I was not looking forward to :grin: as I'm not so familiar with master and there will inevitably be a lot of 4.x specific bugs / ongoing maintenance work.

Particles in particular have the potential to be tricky because (at least in 3.x) they were previously not specified in world space. They (and MultiMesh which CPUParticles used under the hood) were the most time consuming feature in 3.x, though more so for 3D.

Also we should pay particularly attention at whether CPUParticles are planned to be deprecated in 4.x (in which case you will have to get them working with GPUParticles). Also I noted that, at least in 3D, particles in 4.x already use some kind of interpolation so you will need to get both systems to play nice with each other.

Also as I noted above there may end up being some rejigging for snapping purposes. This is something I'm anticipating may need adjustment in 3.x based on feedback, as we haven't yet released 3.6. So in some ways this code may get more testing in 4.x (also bearing in mind there are now many more users on 4.x than 3.x). The paradigm is intended to be that transforms are interpolated and passed through the tree _without snapping_, and snapping is applied on a per object basis. Consider if a child node has a scaling applied, then trying to interpolate from snapped parent transforms would look incorrect otherwise.

Calinou commented:
> Also we should pay particularly attention at whether CPUParticles are planned to be deprecated in 4.x (in which case you will have to get them working with GPUParticles). Also I noted that, at least in 3D, particles in 4.x already use some kind of interpolation so you will need to get both systems to play nice with each other.

Only GPU particles have built-in interpolation in 4.x, and [it's not complete yet](https://github.com/godotengine/godot/issues/51318).

rburing commented:
Thanks for the feedback @lawnjelly ! I think it makes sense to leave particles for a follow-up PR, if reviewers can agree.

I'll have to look at the snapping (propagation) more carefully, and I'll ask for some more eyes on this on RocketChat.

The latest push also fixes this issue with TileMaps:


https://github.com/godotengine/godot/assets/229837/9f511cad-e07c-48ee-8be6-af340bd62975

rburing commented:
The test from the project [PixelPerfectTest.zip](https://github.com/godotengine/godot/files/13244300/PixelPerfectTest.zip) from https://github.com/godotengine/godot/pull/84380 seems to pass with a physics tick rate of 30 (also lower ones) plus the interpolation from this PR, after commenting out the `_process` function in `Camera.gd` (it was setting the camera position in both `_process` and `_physics_process`):

https://github.com/godotengine/godot/assets/229837/956bfec4-04ce-45f9-9ced-eddc8e76dcbf

This was recorded using Peek because movie writer mode is bugged when using the settings of the test project.

The potential issue https://github.com/godotengine/godot/pull/88424#discussion_r1493281555 with propagating snapped transforms to children should be present in `master` as well, so I'd prefer not to change the behavior in this PR. What would an MRP look like @lawnjelly? It can be reported separately.

Mickeon commented on doc/classes/SceneTree.xml:
```diff
@@ -263,6 +263,9 @@
 			- [method Node._process], [method Node._physics_process] and [method Node._input] will not be called anymore in nodes.
 		</member>
+		<member name="physics_interpolation" type="bool" setter="set_physics_interpolation_enabled" getter="is_physics_interpolation_enabled" default="false">
+			Although physics interpolation would normally be globally turned on and off using [member ProjectSettings.physics/common/physics_interpolation], this property allows control over interpolation at runtime.
```
This needs to be entirely reworded but I have no specific suggestion at the moment ([working on something else](https://github.com/godotengine/godot/pull/87334)).
Something more akin to how the equivalent Project Setting is described, which then ends in like "This corresponds to the ... project setting".

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -2262,9 +2262,14 @@
 		<member name="physics/common/physics_jitter_fix" type="float" setter="" getter="" default="0.5">
 			Controls how much physics ticks are synchronized with real time. For 0 or less, the ticks are synchronized. Such values are recommended for network games, where clock synchronization matters. Higher values cause higher deviation of in-game clock and real clock, but allows smoothing out framerate jitters. The default value of 0.5 should be good enough for most; values above 2 could cause the ga ...
-			[b]Note:[/b] For best results, when using a custom physics interpolation solution, the physics jitter fix should be disabled by setting [member physics/common/physics_jitter_fix] to [code]0[/code].
+			[b]Note:[/b] For best results, when using a physics interpolation solution (such as [member physics/common/physics_interpolation] or a custom solution), the physics jitter fix should be disabled by setting [member physics/common/physics_jitter_fix] to [code]0[/code].
```
```suggestion
			[b]Note:[/b] For best results, when using physics interpolation (such as by enabling [member physics/common/physics_interpolation] or using a custom solution), the physics jitter fix should be disabled by setting [member physics/common/physics_jitter_fix] to [code]0[/code].
```

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -2262,9 +2262,14 @@
 		</member>
+		<member name="physics/common/physics_interpolation" type="bool" setter="" getter="" default="false">
+			If [code]true[/code], the renderer will interpolate the transforms of physics objects between the last two transforms, so that smooth motion is seen even when physics ticks do not coincide with rendered frames.
+			[b]Note:[/b] When moving objects to new positions (rather than the usual physics motion) you may want to temporarily turn off interpolation to prevent a visible glitch. You can do this using the [method Node.reset_physics_interpolation] function.
```
"When moving objects" still implies movement in between which is the exact opposite of what is intended, so it feels wrong to say it like this.
Maybe "teleporting" or "shifting" would be more understandable?
```suggestion
			[b]Note:[/b] When shifting objects to completely new positions, you may want to temporarily disable interpolation with [method Node.reset_physics_interpolation].
```

Calinou reviewed (commented):
Tested locally (rebased on top of `master` b8f106acb24387d2f2c746a586edc0728fd10aaf), it works as expected. Amazing work!

Here's the 2D platformer demo with **Physics Ticks per Second** set to `10`:

https://github.com/godotengine/godot/assets/180032/6f80776a-b351-49cd-9146-b2c24e8396d2

___

I noticed an issue when resizing the viewport when using the `viewport` stretch mode and a Camera2D. If the aspect ratio changes, the camera motion will be interpolated, even though this doesn't occur in discrete steps when physics interpolation is disabled. The Camera2D in the project has its **Process Callback** set to **Idle** (the default).

This is hardly noticeable with tickrates greater than 30 Hz, so it's not a huge issue, but it might be worth looking into nonetheless.

See these videos with **Physics Ticks per Second** set to `5`:

### Physics interpolation disabled

https://github.com/godotengine/godot/assets/180032/b0e70667-a47e-4a03-8d1b-a628abe9af30

### Physics interpolation enabled

https://github.com/godotengine/godot/assets/180032/c4c93f0c-7090-465e-b271-b1c2b81f9c45

___

In the bullet shower demo with **Physics Ticks per Second** set to `5`, something strange happens with the sprite movement. It's interpolated even though we're moving it every frame in `_input()` based on mouse movement, so it can move faster than the physics tick rate.

The individual white sprites (that are not nodes but created via low-level servers) are also not interpolated:

### Physics interpolation disabled

https://github.com/godotengine/godot/assets/180032/9ecf5f35-ecec-425e-90be-fdb473905ae4

### Physics interpolation enabled

https://github.com/godotengine/godot/assets/180032/2f6c051e-a910-4692-98d2-9ca4fae7b8f4

___

Parallax backgrounds look strange when wrapping over with low physics tick rate (10 here), and they don't interpolate as smoothly as they could. They seem to interpolate in visible "steps", as if the interpolation was only performed for half of the physics tick length:

https://github.com/godotengine/godot/assets/180032/39102d5b-6b7c-406d-a5a7-9ac8313ea9c1

**Testing project:** [physics_platformer_modified.zip](https://github.com/godotengine/godot/files/14349244/physics_platformer_modified.zip)

___

Also, I strongly suggest enabling physics interpolation by default. We'll need to stick to a default value once this is merged, and with high refresh rate displays slowly becoming the norm among players, it's expected that games should adapt automatically without requiring extra work from the developer's side. While this will require updating some tutorials, I believe it's worth the effort.

rburing commented:
@Calinou Thanks a lot for your testing!

I've rebased the PR and fixed the issue with parallax (it should not be interpolated, as it is already updated every frame).

I'm not sure what to do about the player sprite in the bullet shower demo. The bullets, with their transforms set manually every physics tick, are rendered using `_draw` with the position from the last physics tick, so it is expected that they are not interpolated automagically. They can be interpolated by modifying the `_draw` function e.g. as follows:

```GDScript
func _draw():
    var offset = -bullet_image.get_size() * 0.5
    for bullet in bullets:
        var interpolated_position = lerp(bullet.position + Vector2(bullet.speed / Engine.physics_ticks_per_second, 0), bullet.position, Engine.get_physics_interpolation_fraction())
        draw_texture(bullet_image, interpolated_position + offset)
```

(It could be improved a bit further by storing the current and previous position, thereby avoiding the hack to compute `delta`, and setting both the current and previous position when wrapping around, effectively resetting the interpolation.)

I haven't yet looked into the Camera2D interpolation when resizing the viewport with stretch mode set to `viewport`, but it seems a low priority issue.

It's easy to make physics interpolation the default (should it also be a basic setting then?), but currently this would also override the `Engine.physics_jitter_fix` to `0.0`, which would be good for 2D with interpolation but maybe bad for 3D which is currently lacking interpolation.

Calinou commented:
> It's easy to make physics interpolation the default (should it also be a basic setting then?)

I'm not sure if it should be made a basic setting, considering we want to encourage leaving physics interpolation enabled as much as possible.

lawnjelly commented on scene/2d/camera_2d.cpp:
```diff
@@ -159,8 +173,17 @@ Transform2D Camera2D::get_camera_transform() {
+		// during the call to get_camera_transform().
+		// It may be called MULTIPLE TIMES on certain frames,
+		// therefore smoothing is not currently applied only once per frame / tick,
+		// which will result in some haphazard results.
```
Just to note, this mentioned bug isn't introduced by physics interpolation, it is pre-existing, and this note is to remind to fix it, or make obvious if it appears in a later issue.

Fixing may need some care as `Camera2D` is pretty central to 2D games and there would be risk of regressions.

lawnjelly commented:
Now I've had a look through, it looks pretty straightforward and good to me, and testing seems to be okay so far. The 4.x reviewers might have some preferences for cosmetic changes to naming etc.

One cosmetic thing to pay attention to might be the `physics_interpolation_mode` on Nodes - deciding whether to put it in a tab / group or have it out on its own. It made sense in 3.x, but imo it's easy to bog down the inspector properties with "cognitive overload" (especially for base classes), so hiding things in tabs (groups) can be helpful, and thinking ahead so similarish things can be grouped together in future is usually a good plan.

Mickeon commented on doc/classes/Node.xml:
```diff
@@ -1122,6 +1149,9 @@
 			Notification received when the node is enabled again after being disabled. See [constant PROCESS_MODE_DISABLED].
 		</constant>
+		<constant name="NOTIFICATION_RESET_PHYSICS_INTERPOLATION" value="69">
+			Notification received when [method reset_physics_interpolation] is called on the node or parent nodes.
```
```suggestion
			Notification received when [method reset_physics_interpolation] is called on the node or its ancestors.
```

Mickeon reviewed (commented):
Holy crap that's a whooping 15 comments and I'm not done yet.

rburing commented on doc/classes/Node.xml:
```diff
@@ -1237,6 +1267,15 @@
+			Inherits physics interpolation mode from the node's parent. For the root node, it is equivalent to [constant PHYSICS_INTERPOLATION_MODE_ON]. This is the default for any newly created node.
+		</constant>
+		<constant name="PHYSICS_INTERPOLATION_MODE_OFF" value="1" enum="PhysicsInterpolationMode">
+			Turn off physics interpolation in this node and in children set to [constant PHYSICS_INTERPOLATION_MODE_INHERIT].
```
I see what you mean, but I don't agree with leaving out the mention of children. I think it is good to emphasize that children will be updated. If you leave it out, it is not clear that the children wouldn't be left with some stale value that the parent used to have.

rburing commented:
I fixed the overzealous internal process mode updates when setting a Camera2D to current, and the documentation.

Left to decide:

- [ ] Enable physics interpolation by default or not? (Also considering the jitter fix setting and 3D.)
- [ ] Put the physics interpolation mode property in its own group or not?

Left to do:

- [ ] Get feedback from @godotengine/rendering about the remaining TODOs in the code (particularly about threading).

AttackButton commented:
This could probably fix https://github.com/godotengine/godot/issues/58269 and make sync_to_physics obsolete. Despite the cool name, sync to physics is just a hack, a ugly one.

Calinou commented:
> This could probably fix #58269 and make sync_to_physics obsolete. Despite the cool name, sync to physics is just a hack, a ugly one.

I'm not sure if physics interpolation makes Sync to Physics obsolete. The point of Sync to Physics is ensure that when a player is on a moving platform, the player won't lag behind the platform (and therefore appear to float over it when it moves downwards, or sink into it when it moves upwards). Enabling physics interpolation does not change how physics are calculated; it only changes how physics are *displayed*.

AttackButton commented:
This happened in Godot 3 physics. In version 4, this does not happen anymore. Unless there's another character on top of another or something like that. Either way, this type of thing happens whether sync_to_physics is enabled or not.

Nowadays, this property is only used to synchronize physics and visuals, the physics works the same, not to suggest it might be worse.

The problem with the issue I mentioned isn't with the physics itself, it's because sync_to_physic uses a certain type of notification that doesn't match the one used by PathFollow.

Anyway, I just want to test if this pr can make the physics and visuals synchronize correctly on a moving platform with sync_to_physics disabled and a character on top of it.

AttackButton commented:
> > This could probably fix #58269 and make sync_to_physics obsolete. Despite the cool name, sync to physics is just a hack, a ugly one.
> 
> I'm not sure if physics interpolation makes Sync to Physics obsolete. The point of Sync to Physics is ensure that when a player is on a moving platform, the player won't lag behind the platform by 1 physics frame (and therefore appear to float over it when it moves downwards, or sink into it when it moves upwards). Enabling physics interpolation does not change how physics are calculated; it only changes how physics are _displayed_.

You are right. I was hoping it was just a visual thing.

lawnjelly commented on drivers/gles3/rasterizer_canvas_gles3.cpp:
```diff
@@ -226,7 +226,8 @@ void RasterizerCanvasGLES3::canvas_render_items(RID p_to_render_target, Item *p_
 
-			Vector2 canvas_light_pos = p_canvas_transform.xform(l->xform.get_origin()); //convert light position to canvas coordinates, as all computation is done in canvas coords to avoid precision loss
+			// Convert light position to canvas coordinates, as all computation is done in canvas coordinates to avoid precision loss.
+			Vector2 canvas_light_pos = p_canvas_transform.xform(l->xform_curr.get_origin()); // TODO: Interpolate or explain why not needed.
```
Don't know whether it helps, but in 3.x, the interpolation for canvas light is done in `visual_server_viewport.cpp` line 143:

```
real_t f = Engine::get_singleton()->get_physics_interpolation_fraction();
TransformInterpolator::interpolate_transform_2d(cl->xform_prev, cl->xform_curr, cl->xform_cache, f);
cl->xform_cache = xf * cl->xform_cache;
```

rburing commented:
Thanks @lawnjelly for the comments!

I've left the feature disabled by default, because enabling this on projects which are not made with this in mind can introduce streaking, and because particles and 3D are not done yet. We can revisit the default later. Let's ask people to test this on their 2D games in the release blog post and on social media.

With that I think this PR is ready to be merged, as soon as it can get a review.

lawnjelly commented:
`reset_physics_interpolation()` doesn't appear to be working for `PointLight2D`.
Not sure whether that is also the case in 3.x (it may be my mistake :grin: ).

adamscott reviewed (approved):
What a great port. I tested it summarily (and it seems to work amazingly), but I think others tested it thoroughly.

The code seems OK! Didn't find anything requiring major changes, but this is a hefty PR, so I suggest that we merge for dev6 then we fix issues afterwards.

adamscott commented on scene/2d/camera_2d.cpp:
```diff
@@ -54,7 +54,12 @@ void Camera2D::_update_scroll() {
 		ERR_FAIL_COND(custom_viewport && !ObjectDB::get_instance(custom_viewport_id));
 
-		Transform2D xform = get_camera_transform();
+		Transform2D xform;
+		if (is_physics_interpolated_and_enabled()) {
+			xform = _interpolation_data.xform_prev.interpolate_with(_interpolation_data.xform_curr, Engine::get_singleton()->get_physics_interpolation_fraction());
+		} else {
+			xform = get_camera_transform();
+		}
```
Nevermind, it seems that the compiler can optimize these assignations. (see [this discussion](https://chat.godotengine.org/channel/devel?msg=qcQxkDpG4rDbCFkKa))

lawnjelly commented on servers/rendering/renderer_canvas_cull.h:
```diff
@@ -270,6 +270,10 @@ class RendererCanvasCull {
 
+	void canvas_item_set_interpolated(RID p_item, bool p_interpolated);
+	void canvas_item_reset_physics_interpolation(RID p_item);
+	void canvas_item_transform_physics_interpolation(RID p_item, Transform2D p_transform);
```
Good point, I'll actually fix these in 3.x. I can't remember why I didn't use const ref at the time, perhaps it was just a late night. :grin:

reduz reviewed (approved):
Looks fantastic, really good work!

markdibarry commented:
@rburing From your comment earlier, you probably want to add your new method to `Parallax2D` as well, since (like `ParallaxLayer`) it also relies on immediately jumping between two points for the infinite scroll effect:

```cpp
void Parallax2D::_update_repeat() {
	if (!is_inside_tree()) {
		return;
	}

	Point2 repeat_scale = repeat_size * get_scale();
	RenderingServer::get_singleton()->canvas_set_item_repeat(get_canvas_item(), repeat_scale, repeat_times);
	RenderingServer::get_singleton()->canvas_item_set_interpolated(ci, false);
}
```

rburing commented:
Thanks for the reviews! And thanks @markdibarry, it's done. This should be ready to merge now.

akien-mga commented:
Thanks! Amazing work!

aaronfranke commented on main/main.cpp:
```diff
@@ -3967,6 +3967,11 @@ bool Main::iteration() {
 		PhysicsServer3D::get_singleton()->flush_queries();
 #endif // _3D_DISABLED
 
+		// Prepare the fixed timestep interpolated nodes BEFORE they are updated
+		// by the physics server, otherwise the current and previous transforms
+		// may be the same, and no interpolation takes place.
+		OS::get_singleton()->get_main_loop()->iteration_prepare();
```
If the goal is to do this before physics, why isn't it done before all physics (including 3D physics?)

lawnjelly replied:
@aaronfranke You are also absolutely right, I didn't spot this in the PR, the prepare should probably be before the 3D here (although it shouldn't make any difference until the 3D is done, so we can fix in 3D PR).

To address the confusion a little - "physics interpolation" (somewhat confusingly) is nothing to do with physics engine. The more correct name is fixed timestep interpolation.

The only relation is to the physics ticks, rather than the physics itself. It can work with or without physics engine. Juan preferred the name "physics interpolation", because even though incorrect, he thought it would be easier for users to understand (and admittedly in many cases their eyes would glaze over with "fixed timestep interpolation" :grin: ).

This is something that we as developers will be expected to learn the difference, but it will inevitably cause confusion in some way (if we e.g. called it "fixed timestep interpolation" in the c++, and "physics interpolation" for user facing code, this could be confusing, as many c++ funcs are bound etc, so it is called "physics interpolation" throughout).

As it is not related to the physics engines, it should not be disabled when disabling physics engines (as `_physics_process()` still occurs).

Just to try and explain what it is doing .. the `VisualServer` items have to be prepared before each iteration (tick) before any changes have been made to them (server side at least).

If you follow the code it calls `update_interpolation_tick()`.

This does a bunch of things:
* Detect any objects that are no longer being moved, and ensure their previous transform = current transform so they cease vibrating
* Store the current transform (from the last tick) to the previous transform, so they can be interpolated after the tick
